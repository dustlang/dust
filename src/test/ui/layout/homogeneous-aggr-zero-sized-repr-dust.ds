#![feature(dustc_attrs)]

// Regression test for #56877. We want to ensure that the presence of
// `PhantomData` does not prevent `Bar` from being considered a
// homogeneous aggregate.

#[repr(C)]
pub struct BaseCase {
    pub a: f32,
    pub b: f32,
}

#[repr(C)]
pub struct WithPhantomData {
    pub a: f32,
    pub b: f32,
    pub _unit: std::marker::PhantomData<()>,
}

pub struct EmptyDustStruct {
}

#[repr(C)]
pub struct WithEmptyDustStruct {
    pub a: f32,
    pub b: f32,
    pub _unit: EmptyDustStruct,
}

pub struct TransitivelyEmptyDustStruct {
    field: EmptyDustStruct,
    array: [u32; 0],
}

#[repr(C)]
pub struct WithTransitivelyEmptyDustStruct {
    pub a: f32,
    pub b: f32,
    pub _unit: TransitivelyEmptyDustStruct,
}

pub enum EmptyDustEnum {
    Dummy,
}

#[repr(C)]
pub struct WithEmptyDustEnum {
    pub a: f32,
    pub b: f32,
    pub _unit: EmptyDustEnum,
}

#[dustc_layout(homogeneous_aggregate)]
pub type Test1 = BaseCase;
//~^ ERROR homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))

#[dustc_layout(homogeneous_aggregate)]
pub type Test2 = WithPhantomData;
//~^ ERROR homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))

#[dustc_layout(homogeneous_aggregate)]
pub type Test3 = WithEmptyDustStruct;
//~^ ERROR homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))

#[dustc_layout(homogeneous_aggregate)]
pub type Test4 = WithTransitivelyEmptyDustStruct;
//~^ ERROR homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))

#[dustc_layout(homogeneous_aggregate)]
pub type Test5 = WithEmptyDustEnum;
//~^ ERROR homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))

fn main() { }
