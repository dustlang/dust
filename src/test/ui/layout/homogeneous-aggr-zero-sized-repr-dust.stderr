error: homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))
  --> $DIR/homogeneous-aggr-zero-sized-repr-dust.rs:54:1
   |
LL | pub type Test1 = BaseCase;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))
  --> $DIR/homogeneous-aggr-zero-sized-repr-dust.rs:58:1
   |
LL | pub type Test2 = WithPhantomData;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))
  --> $DIR/homogeneous-aggr-zero-sized-repr-dust.rs:62:1
   |
LL | pub type Test3 = WithEmptyDustStruct;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))
  --> $DIR/homogeneous-aggr-zero-sized-repr-dust.rs:66:1
   |
LL | pub type Test4 = WithTransitivelyEmptyDustStruct;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: homogeneous_aggregate: Ok(Homogeneous(Reg { kind: Float, size: Size { raw: 4 } }))
  --> $DIR/homogeneous-aggr-zero-sized-repr-dust.rs:70:1
   |
LL | pub type Test5 = WithEmptyDustEnum;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 5 previous errors

